src/lib/group.ts

```diff
@@ -0,0 +1,364 @@
+   */
+  neg() {
+    let { x, y } = this;
+    return new Group({ x, y: y.neg() });
+  }
+
+  /**
+   * Elliptic curve scalar multiplication. Scales the {@link Group} element `n`-times by itself, where `n` is the {@link Scalar}.
+   *
+   * ```typescript
+   * let s = Scalar(5);
+   * let 5g = g.scale(s);
+   * ```
+   */
+  scale(s: Scalar | number | bigint) {
+    let scalar =
+      typeof s === 'bigint' || typeof s === 'number'
+        ? Scalar.fromBigInt(BigInt(s))
+        : s;
+    let fields = scalar.toFields();
+
+    if (this.#isConstant() && fields.every((f) => f.isConstant())) {
+      let g_proj = Pallas.scale(this.#toAffine(), BigInt(scalar.toJSON()));
+      return Group.#fromProjective(g_proj);
+    } else {
+      let [, x, y] = Snarky.group.scale(this.#toTuple(), [
+        0,
+        ...fields.map((f) => f.value).reverse(),
+      ]);
+      return new Group({ x, y });
+    }
+  }
+
+  /**
+   * Assert that this {@link Group} element equals another {@link Group} element.
+   * Throws an error if the assertion fails.
+   *
+   * ```ts
+   * g1.assertEquals(g2);
+   * ```
+   */
+  assertEquals(g: Group, message?: string) {
+    let { x: x1, y: y1 } = this;
+    let { x: x2, y: y2 } = g;
+
+    x1.assertEquals(x2, message);
+    // y1.assertEquals(y2, message); need to enable this later on, but it breaks constraint backwards compatibility
+  }
+
+  /**
+   * Check if this {@link Group} element equals another {@link Group} element.
+   * Returns a {@link Bool}.
+   *
+   * ```ts
+   * g1.equals(g1); // Bool(true)
+   * ```
+   */
+  equals(g: Group) {
+    if (this.#isConstant() && g.#isConstant()) {
+      let { x: x1, y: y1 } = this;
+      let { x: x2, y: y2 } = g;
+
+      return Bool(x1.equals(x2).and(y1.equals(y2)));
+    } else {
+      /*
+    let equal_x = x1.equals(x2);
+    let equal_y = y1.equals(y2);
+    return equal_x.and(equal_y);*/
+
+      let z = Snarky.group.equals(this.#toTuple(), g.#toTuple());
+      return Bool.Unsafe.ofField(new Field(z));
+    }
+  }
+
+  /**
+   * Serializes this {@link Group} element to a JSON object.
+   *
+   * This operation does NOT affect the circuit and can't be used to prove anything about the representation of the element.
+   */
+  toJSON(): {
+    x: string;
+    y: string;
+  } {
+    return {
+      x: this.x.toString(),
+      y: this.y.toString(),
+    };
+  }
+
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Returns an array containing this {@link Group} element as an array of {@link Field} elements.
+   */
+  toFields() {
+    return [this.x, this.y];
+  }
+
+  /**
+   * Coerces two x and y coordinates into a {@link Group} element.
+   */
+  static from(
+    x: FieldVar | Field | number | string | bigint,
+    y: FieldVar | Field | number | string | bigint
+  ) {
+    return new Group({ x, y });
+  }
+
+  /**
+   * Adds a {@link Group} element to another one.
+   */
+  static add(g1: Group, g2: Group) {
```

Comment:
IMO all these extra static methods are weird and they should either be deprecated or be present consistently on all classes, i.e. added to `Field` for example

Reply:
I am also not a massive fan, I would rather deprecate them

Reply:
We can always recreate them with `bind` anyway -- JS devs are used to this; so I agree we should deprecate